### PR TITLE
Copy all files from web folder to build output

### LINKED ISSUE #42716: [web] Control asset path for service workers
## Use case
For service workers on web it's important that the get installed at the root. If the get registered somewhere under `/assets/...` they can't work in the intended way.

## Proposal
The easiest solution would be to copy the whole content of `web/` to `build/web/`, but there could be some problems with that solution.

### PR DESCRIPTION
**Pull request by provokateurin**
## Description
Copies all files from the `web` folder to the `build/web` folder. Needed for service workers to function correctly. They need to be under the root location.

## Related Issues

Fixes #42716

## Tests

I added the following tests:

There were no tests that tested if the `index.html` has been copied correctly so I was not sure were to add the test and how the test should look like. Please give me advice and I'll fix that.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jld3103** [comment]
@jonahwilliams Can you help me?

**Comment by jonahwilliams** [comment]
You'll need to add a test that this behaves as expected. There are some existing tests here https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/general.shard/build_system/targets/web_test.dart .

The test case should include several arbitrary files in the web directory, execute the target, and then assert they are contained in the ouput.

Additionally, we need to ensure that cleaning the target works correctly. This requires you to tell the build system about what files you actually copied. Check the Depfile class a few lines below. You'll need to add an additional depfile that contains all of the copied files.

**Comment by jld3103** [comment]
@jonahwilliams Sorry but I'm not sure how to create the test.
This is my test code:
```dart
  test('WebReleaseBundle copies all files from web folder', () => testbed.run(() async {
      fs.file(fs.path.join('lib', 'main.dart'))
          ..createSync(recursive: true)
        ..writeAsStringSync('hello');
      fs.file(fs.path.join('web', 'index.html'))
        ..createSync(recursive: true)
        ..writeAsStringSync('hello');
      fs.file(fs.path.join('web', 'test.js'))
        ..createSync(recursive: true)
        ..writeAsStringSync('alert(\'hello world\');');
      fs.file(fs.path.join('web', 'abc.txt'))
        ..createSync(recursive: true)
        ..writeAsStringSync('hello world');
      await const WebReleaseBundle().build(environment);

      expect(environment.buildDir.childFile('test.js').existsSync(), true);
      expect(environment.buildDir.childFile('test.js').readAsStringSync(),'alert(\'hello world\');');
      expect(environment.buildDir.childFile('abc.txt').existsSync(), true);
      expect(environment.buildDir.childFile('abc.txt').readAsStringSync(),'hello world');
    }, overrides: <Type, Generator>{
    ProcessManager: () => MockProcessManager(),
  }));
```
The test fails with this error:
```
  NoSuchMethodError: The method '[]' was called on null.
  Receiver: null
  Tried calling: []("HOME")
  dart:core                                                         Object.noSuchMethod
  package:flutter_tools/src/base/file_system.dart 194:30            userHomePath
  package:flutter_tools/src/base/config.dart 15:54                  new Config
  package:flutter_tools/src/context_runner.dart 83:21               runInContext.<fn>
  package:flutter_tools/src/base/context.dart 100:41                AppContext._generateIfNecessary.<fn>
  dart:collection                                                   _LinkedHashMapMixin.putIfAbsent
  package:flutter_tools/src/base/context.dart 88:20                 AppContext._generateIfNecessary
  package:flutter_tools/src/base/context.dart 117:32                AppContext.get
  package:flutter_tools/src/base/context.dart 115:23                AppContext.get
  package:flutter_tools/src/base/config.dart 31:41                  Config.instance
  package:flutter_tools/src/globals.dart 14:29                      config
  package:flutter_tools/src/build_info.dart 525:26                  getBuildDirectory
  package:flutter_tools/src/build_info.dart 550:23                  getAssetBuildDirectory
  package:flutter_tools/src/asset.dart 119:22                       _ManifestAssetBundle.build
  package:flutter_tools/src/build_system/targets/assets.dart 22:21  copyAssets
  package:flutter_tools/src/build_system/targets/web.dart 218:35    WebReleaseBundle.build
  test/general.shard/build_system/targets/web_test.dart 327:38      main.<fn>.<fn>
  test/src/testbed.dart 139:23                                      Testbed.run.<fn>.<fn>.<fn>
 
```

**Comment by jonahwilliams** [comment]
you'll need to provide a platform mock which has a home directory set, something like:

```
when(platform.environment).thenReturn(<String, String>{});
...
Platform: () => mockPlatform()
```

If you push the partially working code I can take a look at at too

**Comment by efichot** [comment]
Waiting for the merge 🙃

**Comment by jld3103** [comment]
@jonahwilliams sorry for the long delay.

**Comment by jonahwilliams** [comment]
This doesn't add the copied files to the depfile, so deletion does not work as expected. Here is a testcase for this:

```

 test('WebReleaseBundle copies all files from web folder', () => testbed.run(() async {
    fs.file(fs.path.join('web', 'index.html'))
      ..createSync(recursive: true)
      ..writeAsStringSync('hello');
    final File textFile = fs.file(fs.path.join('web', 'abc.txt'))
      ..createSync(recursive: true)
      ..writeAsStringSync('hello world');
    await const WebReleaseBundle().build(environment);

    expect(environment.outputDir.childFile('abc.txt').existsSync(), true);

    textFile.deleteSync();
    await const WebReleaseBundle().build(environment);

   expect(environment.outputDir.childFile('abc.txt').existsSync(), false);
  }, overrides: <Type, Generator>{
    ProcessManager: () => MockProcessManager(),
    Platform: () => FakePlatform()..environment = <String, String>{'HOME': '/home/me'},
  }));
```

File: packages/flutter_tools/test/general.shard/build_system/targets/web_test.dart [hunk 1]
```diff
@@ -310,6 +310,27 @@ void main() {
...
+      ..writeAsStringSync('hello world');
+    await const WebReleaseBundle().build(environment);
+
+    expect(environment.outputDir.childFile('test.js').existsSync(), true);
+    expect(environment.outputDir.childFile('test.js').readAsStringSync(),'alert(\'hello world\');');
+    expect(environment.outputDir.childFile('abc.txt').existsSync(), true);
+    expect(environment.outputDir.childFile('abc.txt').readAsStringSync(),'hello world');
+  }, overrides: <Type, Generator>{
+    ProcessManager: () => MockProcessManager(),
+    Platform: () => FakePlatform()..environment = <String, String>{'HOME': '/home/me'},
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/build_system/targets/web_test.dart]: This value seems overly specific. What config is depending on this? Could that be mocked instead?

Comment (jld3103) [reply packages/flutter_tools/test/general.shard/build_system/targets/web_test.dart]: Which value do you exactly mean? The home directory? I took that one from some other tests for android. But I guess you mean something else?

**Comment by jld3103** [comment]
@jonahwilliams Actually I was not able to find any test that tests if the files get cleaned correctly after rebuilding. I'll push my changes and let you have a look at it.

**Comment by jld3103** [comment]
@jonahwilliams ping

File: packages/flutter_tools/lib/src/build_system/targets/web.dart [hunk 2]
```diff
@@ -214,11 +215,53 @@ class WebReleaseBundle extends Target {
...
-      .copySync(fs.path.join(environment.outputDir.path, 'index.html'));
-    final Depfile depfile = await copyAssets(environment, environment.outputDir.childDirectory('assets'));
-    depfile.writeToFile(environment.buildDir.childFile('flutter_assets.d'));
+    final String basePath = environment.projectDir.childDirectory('web').absolute.path;
+    final List<File> inputs = <File>[];
+    final List<File> outputs = <File>[];
+    final String outputPrefix = fs.path.join(
+      environment.projectDir.path,
+    );
+    for (String artifact in environment.projectDir.childDirectory('web').listSync(recursive: true).map((FileSystemEntity file) => file.absolute.path)) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/web.dart]: recursive: true will give you nested files with the correct path. you could probably update this to:

`final List<File> files = environment.projectDir.childDirectory('web').listSync(recursive: true).whereType<File>().toList();`

This functionality could also be pulled out into a helper function to make it easier to read

File: packages/flutter_tools/test/general.shard/build_system/targets/web_test.dart [hunk 3]
```diff
@@ -310,6 +310,26 @@ void main() {
   }, overrides: <Type, Generator>{
     ProcessManager: () => MockProcessManager(),
   }));
+
+  test('WebReleaseBundle copies all files from web folder', () => testbed.run(() async {
+    fs.file(fs.path.join('web', 'index.html'))
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/build_system/targets/web_test.dart]: You should add a file in a directory like web/foo/bar.js to test that it gets copied to the correct path

**Comment by zanderso** [comment]
@jonahwilliams is this PR superseded by one of your recent changes?

**Comment by jonahwilliams** [comment]
This PR is superseded by https://github.com/flutter/flutter/pull/48316

**Pull request closed without merging**